### PR TITLE
Mvp/notifications (Questions and Answers)

### LINKED ISSUE #13: Notification system
**Description**
- Implement notifications storage and creation when triggered. 
- Send email push notifications when needed.

**Requirements**

- [x] [Blocks everything] Create notification entity (row and columns)
- [x] [Depends on setup of MySql database] Fetch notifications related to an id to populate the inbox dropdown.
- [x] Create notification when question answered (mentee).
- [x] Create notification when answer is commented (mentor).
- [ ] Create notification when meeting is scheduled (mentor).
- [ ] Create notification when new mentor applies for sign up (mentor).
- [ ] [OPTIONAL] Trigger notification when chat is requested (mentor).

### PR DESCRIPTION
**Pull request by abundis28**
# What does this PR do?
- Enables Notifications for questions and answers.

**Related Issues**
- Closes #13 

---
**Merge Checklist**
- [X] Branch is up to date
- [ ] No code style issues

### CODE REVIEW

File: src/main/webapp/script.js [hunk 1]
```diff
@@ -12,12 +12,53 @@
...
+  });
+}
+
+/**
+ * Appends child to navbar dropdown. Represents a notification.
+ * @param {Notification} notification
+ */
+function createListElement(notification) {
+  const liElement = document.createElement('li');
+  // Creates a link to redirect the user to the question that was answered or commented.
```
Comment (OUMontiel) [inline src/main/webapp/script.js:34]: nit: Create. Implementation comments should start with imperative verb.

File: src/main/webapp/script.js [hunk 2]
```diff
@@ -12,12 +12,53 @@
...
+  const linkElement = document.createElement('a');
+  linkElement.innerText = linkElement.innerText.concat(notification.message, " - ");
+  linkElement.innerText = linkElement.innerText.concat(notification.timestamp.toString());
+  linkElement.setAttribute("href", notification.url);
+  liElement.appendChild(linkElement);
+  liElement.setAttribute("class","list-group-item");
```
Comment (OUMontiel) [inline src/main/webapp/script.js:40]: You could group better by having the `liElement` declaration later.
```suggestion
  // Create a link to redirect the user to the question that was answered or commented.
  const linkElement = document.createElement('a');
  linkElement.innerText = linkElement.innerText.concat(notification.message, " - ");
  linkElement.innerText = linkElement.innerText.concat(notification.timestamp.toString());
  linkElement.setAttribute("href", notification.url);
  
  // Append link to list element.
  const liElement = document.createElement('li');
  liElement.appendChild(linkElement);
  liElement.setAttribute("class","list-group-item");
```

File: src/main/webapp/script.js [hunk 3]
```diff
@@ -12,12 +12,53 @@
...
+      notificationsElement.appendChild(createListElement(notification));
+    }
+  });
+}
+
+/**
+ * Appends child to navbar dropdown. Represents a notification.
+ * @param {Notification} notification
+ */
+function createListElement(notification) {
```
Comment (OUMontiel) [inline src/main/webapp/script.js:32]: Could you change the function name to be more specific towards notifications (like `createNotificationListElement`)?

File: src/main/webapp/script.js [hunk 4]
```diff
@@ -100,7 +141,6 @@ function fetchAuthentication() {
         // If logged in user is not registered, redirect to signup page.
         window.location.replace(user.authenticationUrl);
       }
-
```
Comment (OUMontiel) [inline src/main/webapp/script.js:103]: Why did you delete this line? (Not asking you to add it back, all I want to know is if there is a guideline I must follow for blank lines.)

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+import javax.servlet.http.HttpServletResponse;
+ 
+/**
+* Servlet that handles the fetching and posting of notifications.
+*/
+@WebServlet("/notification")
+public class NotificationServlet extends HttpServlet {
+  /**
+  * Fetch the notifications of a specific user.
+  */
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:39]: nit: Indent one space.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get user's ID from Utility method.
+    int userId = Utility.getUserId();
+    // If the user is signed in then fetch notifications.
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:44]: I think this comment is not needed (it's repeated below if statement).

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    if (userId != -1) {
+      // If the user is logged in, fetch notifications and convert the ArrayList to JSON
+      // using Utility method.
+      String json = Utility.convertToJsonUsingGson(getNotifications(userId));
+      response.setContentType("application/json;");
+      response.getWriter().println(json);
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:50]: You could save yourself the String declaration:
```suggestion
      response.setContentType("application/json;");
      response.getWriter().println(Utility.convertToJsonUsingGson(getNotifications(userId))); 
```

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+      // using Utility method.
+      String json = Utility.convertToJsonUsingGson(getNotifications(userId));
+      response.setContentType("application/json;");
+      response.getWriter().println(json); 
+    }
+  }
+
+  /**
+  * Post a new notification for all followers of a modified question or answer.
+  */
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:56]: nit#1: Indent one space.
nit#2: Posts. (Javadoc should start with verb in simple present tense. Different from implementation comments.)

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Define local time for the new entries in the server.
+    LocalDateTime localDateTime = LocalDateTime.now();
+    Timestamp localTimestamp = Timestamp.valueOf(localDateTime);
+
+    // Get if notification is question answered or answered commented and the id of the element.
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:63]: Could you rephrase? I'm a bit confused by the wording...

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+    String typeOfNotification = request.getParameter("type");
+    int modifiedElementId = Integer.parseInt(request.getParameter("modifiedElementId"));
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:65]: Isn't this supposed to be `elementId` like in the function notify() from the JS file?

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+    
+    if(typeOfNotification.equals("question")) {
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:67]: nit: Add space before opening parenthesis.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+      // If the notification is for an anwer to a question.
+      String query =  "SELECT follower_id FROM QuestionFollower WHERE question_id = " +
+                       modifiedElementId;
+      // Query the information from QuestionFollower table.
+      try (Connection connection = DriverManager.getConnection(Utility.SQL_LOCAL_URL, 
+                                                               Utility.SQL_LOCAL_USER,
+                                                               Utility.SQL_LOCAL_PASSWORD);
+          PreparedStatement pst = connection.prepareStatement(query);
+          ResultSet rs = pst.executeQuery()) {
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:76]: nit: Indent back two spaces.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+        // Defines the url to which the user will be redirected. Will be defined each case because
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:77]: nit: Capitalize URL.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+        // URL will change once notifications for approvals and meetings are enabled.
+        String elementUrl = "/questions.html?id=" + modifiedElementId;
+        // Insert notification and get its id to relate in UserNotification table.
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:80]: nit: Capitalize ID.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+        int notificationId = getNotificationId(connection, localTimestamp);
+        // Iterate through the query's result set to insert all notifications.
+        while (rs.next()) {
+          insertToUserNotification(connection, rs.getInt(1), notificationId);
+        }
+        // Close the connection once all insertions have been performed.
+        connection.close();
+      } catch (SQLException ex) {
+          Logger lgr = Logger.getLogger(DataServlet.class.getName());
+          lgr.log(Level.SEVERE, ex.getMessage(), ex);
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:118]: I think we should be consistent across files, so could you name these `logger`?

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+                             localTimestamp);
+        int notificationId = getNotificationId(connection, localTimestamp);
+        // Iterate through the query's result set to insert all notifications.
+        while (rs.next()) {
+          insertToUserNotification(connection, rs.getInt(1), notificationId);
+        }
+        // Close the connection once all insertions have been performed.
+        connection.close();
+      } catch (SQLException ex) {
+          Logger lgr = Logger.getLogger(DataServlet.class.getName());
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:117]: Change DataServlet to NotificationServlet.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+          lgr.log(Level.SEVERE, ex.getMessage(), ex);
+      }
+    }
+  }
+
+  /**
+  * Fetches notifications with the user id.
+  */
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:125]: nit#1: Indent one space.
nit#2: Capitalize ID.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+    String query =  "SELECT message, url, date_time FROM Notification WHERE id IN " +
+                    "(SELECT notification_id FROM UserNotification WHERE user_id = " + userId +
+                    ") ORDER BY date_time DESC";
+    List<Notification> notifications = new ArrayList<>();
+    // Query the information from tables and create notification object to be stored in ArrayList.
+    try (Connection connection = DriverManager.getConnection(Utility.SQL_LOCAL_URL,
+                                                             Utility.SQL_LOCAL_USER,
+                                                             Utility.SQL_LOCAL_PASSWORD);
+        PreparedStatement pst = connection.prepareStatement(query);
+        ResultSet rs = pst.executeQuery()) {
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:138]: nit: Indent back two spaces.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+        notification.message = rs.getString(1);
+        notification.url = rs.getString(2);
+        notification.timestamp = rs.getTimestamp(3);
+        // Stores object in ArrayList.
+        notifications.add(notification);
+      }
+      connection.close();
+    } catch (SQLException ex) {
+        Logger lgr = Logger.getLogger(DataServlet.class.getName());
+        lgr.log(Level.SEVERE, ex.getMessage(), ex);
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:151]: Same comment about consistency across files.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+    }
+    return notifications;
+  }
+
+  /**
+  * Function that receives the data for a notification and inserts it into the Notification table.
+  */
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:158]: nit: Indent one space.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 21]
```diff
@@ -0,0 +1,209 @@
...
+  private void insertToNotification(Connection connection, String message, String elementUrl,
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:159]: Delete `connection` variable if you are not using it inside the function.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 22]
```diff
@@ -0,0 +1,209 @@
...
+      prepStatement.executeUpdate();
+    } catch (SQLException e) {
+      System.out.println(e.getMessage());
+    }
+  }
+
+  /**
+  * Function that receives the data for a notification-user relation and inserts it into
+  * UserNotification.
+  */
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:177]: nit: Indent one space.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 23]
```diff
@@ -0,0 +1,209 @@
...
+  private void insertToUserNotification(Connection connection, int userId, int notificationId) {
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:178]: Delete `connection` variable if you are not using it inside the function.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 24]
```diff
@@ -0,0 +1,209 @@
...
+      prepStatement.setInt(2, notificationId);
+      prepStatement.executeUpdate();
+    } catch (SQLException e) {
+      System.out.println(e.getMessage());
+    }
+  }
+
+  /**
+  * Return the id of a just inserted notification.
+  */
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:193]: nit#1: Indent one space.
nit#2: Capitalize ID.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 25]
```diff
@@ -0,0 +1,209 @@
...
+  private int getNotificationId(Connection connection, Timestamp date) {
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:194]: Delete `connection` variable if you are not using it inside the function.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 26]
```diff
@@ -0,0 +1,209 @@
...
+    String query =  "SELECT id FROM Notification ORDER BY date_time";
+    // Query the information from Notification table.
+    int notificationId = 0;
+    try (PreparedStatement pst = connection.prepareStatement(query);
+          ResultSet rs = pst.executeQuery()) {
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:199]: nit: Indent back two spaces.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 27]
```diff
@@ -0,0 +1,209 @@
...
+      // Select the id of the last notification to be inserted.
+      rs.last();
+      notificationId = rs.getInt(1);
+    } catch (SQLException ex) {
+      Logger lgr = Logger.getLogger(DataServlet.class.getName());
+      lgr.log(Level.SEVERE, ex.getMessage(), ex);
```
Comment (OUMontiel) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:205]: Same comment about consistency across files.

File: src/main/webapp/script.js (hunk 1 above)
Comment (abundis28) [reply src/main/webapp/script.js:34]: Ack!

File: src/main/webapp/script.js (hunk 2 above)
Comment (abundis28) [reply src/main/webapp/script.js]: Done!

File: src/main/webapp/script.js (hunk 3 above)
Comment (abundis28) [reply src/main/webapp/script.js:32]: Ack!

File: src/main/webapp/script.js (hunk 4 above)
Comment (abundis28) [reply src/main/webapp/script.js:103]: Mmmm I don't remember erasing it on purpose. I usually don't leave blanks for file size motives but I usually dont mess with other people's code, sorry!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 5 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:39]: Done!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 6 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:44]: Done!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 7 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:50]: Done!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 8 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:56]: Done!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 9 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:63]: Ack

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 10 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:65]: Fixed!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 11 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:67]: Done!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 12 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:76]: Indented it forward one space to align with connection.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 13 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java]: Done!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 14 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:80]: done

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 15 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:118]: done!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 16 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:117]: Done!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 17 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:125]: Done!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 18 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:138]: Done!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 21 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:159]: It is being used.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 23 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:178]: It is being used.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 24 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:193]: Done

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 25 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:194]: It is being used.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 26 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:199]: Done!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 23 above)
Comment (OUMontiel) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:177]: Oh, right! Didn't see it.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 25 above)
Comment (OUMontiel) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:193]: 👍

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 12 above)
Comment (OUMontiel) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java]: Oh, I just realized you made these lines part of the parenthesis, so yeah, you could just leave the four spaces or align with `connection`.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 21 above)
Comment (OUMontiel) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:158]: 👍

File: src/main/java/com/google/sps/classes/Notification.java [hunk 28]
```diff
@@ -0,0 +1,23 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.classes;
+
+import java.sql.Timestamp;
+
```
Comment (timothy-parsons) [inline src/main/java/com/google/sps/classes/Notification.java:18]: Add a class-level comment?

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 29]
```diff
@@ -0,0 +1,208 @@
...
+@WebServlet("/notification")
+public class NotificationServlet extends HttpServlet {
+  /**
+   * Fetches the notifications of a specific user.
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get user's ID from Utility method.
+    int userId = Utility.getUserId();
+    if (userId != -1) {
```
Comment (timothy-parsons) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:44]: I would suggest adding a constant to Constants.java (started by Shaar in https://github.com/abundis28/mintern/pull/32) that is something like Constants.NOT_LOGGED_IN_USER_ID or something similar, so it's clear and easy to refer in multiple places.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 30]
```diff
@@ -0,0 +1,208 @@
...
+      response.setContentType("application/json;");
+      response.getWriter().println(Utility.convertToJsonUsingGson(getNotifications(userId));); 
+    }
+  }
+
+  /**
+   * Posts a new notification for all followers of a modified question or answer.
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (timothy-parsons) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:56]: This is a very very long function. Think through how to split it up into some helper functions! Note that helper functions can either be private methods in this file, or can go in Utility if they will be or could be utilized in multiple files.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 31]
```diff
@@ -0,0 +1,208 @@
...
+    String query = "INSERT INTO Notification(message, url, date_time) VALUES(?,?,?)";
+    try {
+      // Prepare the statement to be inserted.
+      PreparedStatement prepStatement = connection.prepareStatement(query);
+      prepStatement.setString(1, message);
+      prepStatement.setString(2, elementUrl);
+      prepStatement.setTimestamp(3, dateTime);
+      prepStatement.executeUpdate();
+    } catch (SQLException e) {
+      System.out.println(e.getMessage());
```
Comment (timothy-parsons) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:169]: You should log this more formally.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 32]
```diff
@@ -0,0 +1,208 @@
...
+  private void insertToUserNotification(Connection connection, int userId, int notificationId) {
+    String query = "INSERT INTO UserNotification(user_id, notification_id) VALUES(?,?)";
+    try {
+      // Prepare the statement to be inserted.
+      PreparedStatement prepStatement = connection.prepareStatement(query);
+      prepStatement.setInt(1, userId);
+      prepStatement.setInt(2, notificationId);
+      prepStatement.executeUpdate();
+    } catch (SQLException e) {
+      System.out.println(e.getMessage());
```
Comment (timothy-parsons) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:186]: You should log this more formally.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 33]
```diff
@@ -0,0 +1,208 @@
...
+    }
+  }
+
+  /**
+   * Returns the ID of a just inserted notification.
+   */
+  private int getNotificationId(Connection connection, Timestamp date) {
```
Comment (timothy-parsons) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:193]: Maybe change name to 'getLastInsertedNotificationId'

Comment (timothy-parsons) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:193]: The 'date' parameter doesn't seem to be used?

File: src/main/java/com/google/sps/classes/Notification.java (hunk 28 above)
Comment (abundis28) [reply src/main/java/com/google/sps/classes/Notification.java:18]: Done!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 30 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:56]: You are right! I broke up logic inside of if...else... block. Also separated the insertion into helper function. Can you see anyway of further simplifying this code.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 33 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:193]: Ack!

Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:193]: Erased!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 29 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:44]: I erased this part because now the notifications will be loaded only when the user clicks on the inbox button and the button appears only if the user is logged in.

Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:44]: Still I will keep in mind this constants if I come to this situation again.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 31 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java]: Ack!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 32 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java]: Ack!

File: src/main/java/com/google/sps/classes/Notification.java [hunk 34]
```diff
@@ -0,0 +1,26 @@
...
+/**
+ * Class to represent notifications and sent them to front end as converted JSON.
```
Comment (timothy-parsons) [inline src/main/java/com/google/sps/classes/Notification.java:20]: nit: send

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 35]
```diff
@@ -0,0 +1,194 @@
...
+    String query =  "SELECT message, url, date_time FROM Notification WHERE id IN " +
+                    "(SELECT notification_id FROM UserNotification WHERE user_id = " + userId +
+                    ") ORDER BY date_time DESC";
+    List<Notification> notifications = new ArrayList<>();
+    // Query the information from tables and create notification object to be stored in ArrayList.
+    try (Connection connection = DriverManager.getConnection(Utility.SQL_LOCAL_URL,
+                                                             Utility.SQL_LOCAL_USER,
+                                                             Utility.SQL_LOCAL_PASSWORD);
+         PreparedStatement pst = connection.prepareStatement(query);
+         ResultSet rs = pst.executeQuery()) {
```
Comment (timothy-parsons) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:95]: Prefer variable names that are more descriptive - for instance, if I see 'rs' later in the method, I have no idea what it is if I don't remember its initialization.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 36]
```diff
@@ -0,0 +1,194 @@
...
+      // Iterate through the result of the query to populate the ArrayList and return it as JSON.
+      while (rs.next()){
+        Notification notification = new Notification();
+        notification.message = rs.getString(1);
+        notification.url = rs.getString(2);
+        notification.timestamp = rs.getTimestamp(3);
```
Comment (timothy-parsons) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:101]: Use constants here rather than integers. I think it was Omar's PR that I mentioned this on, but if you use a constant like 'rs.getString(SqlConstants.NOTIFICATION_MESSAGE_INDEX)' it will be 1) more readable, and 2) harder to make mistakes and introduce bugs.

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 30 above)
Comment (timothy-parsons) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:54]: This is much better, thank you! Try to keep your functions at a max ~40 lines, though complicated functions can of course get longer, so it's not a hard rule.

File: src/main/java/com/google/sps/classes/Notification.java (hunk 34 above)
Comment (abundis28) [reply src/main/java/com/google/sps/classes/Notification.java:20]: Ack

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 36 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java]: Done!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 35 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java:96]: done!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 37]
```diff
@@ -0,0 +1,195 @@
...
+   */
+  private int getLastInsertedNotificationId(Connection connection) {
+    String query =  "SELECT id FROM Notification ORDER BY date_time";
+    // Query the information from Notification table.
+    int notificationId = 0;
+    try (PreparedStatement pst = connection.prepareStatement(query);
+         ResultSet resultSet = pst.executeQuery()) {
+      // Select the id of the last notification to be inserted.
+      resultSet.last();
+      notificationId = resultSet.getInt(1);
```
Comment (timothy-parsons) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:162]: nit: can you make this '1' into a constant?

File: src/main/java/com/google/sps/servlets/NotificationServlet.java [hunk 38]
```diff
@@ -0,0 +1,195 @@
...
+    try (Connection connection = DriverManager.getConnection(Utility.SQL_LOCAL_URL, 
+            Utility.SQL_LOCAL_USER, Utility.SQL_LOCAL_PASSWORD);
+          PreparedStatement pst = connection.prepareStatement(query);
+          ResultSet resultSet = pst.executeQuery()) {
+      // Insert notification and get its ID to relate in UserNotification table.
+      insertToNotification(connection, notificationMessage, notificationUrl, localTimestamp);
+      int notificationId = getLastInsertedNotificationId(connection);
+      // Iterate through the query's result set to insert all notifications.
+      while (resultSet.next()) {
+        insertToUserNotification(connection, resultSet.getInt(1), notificationId);
```
Comment (timothy-parsons) [inline src/main/java/com/google/sps/servlets/NotificationServlet.java:186]: nit: can you make this '1' into a constant?

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 37 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java]: Done!

File: src/main/java/com/google/sps/servlets/NotificationServlet.java (hunk 38 above)
Comment (abundis28) [reply src/main/java/com/google/sps/servlets/NotificationServlet.java]: Done!
